livecheck migration: add `brew livecheck` developer command

nandahkrishna opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR integrates `livecheck` into Homebrew/brew as a developer command. It was almost just a copy-paste of files, but there were some style changes and a couple of minor differences as outlined by @samford. I've also brought over the command completions and placed them in the appropriate files (works on `zsh`, should also work on `bash` and `fish` based on my previous trials).

I tested this locally by also incorporating the changes in #8156 (without which we will get an `undefined method strategy` error), and it works as intended (I also untapped homebrew/livecheck).

Things to discuss:
1. What changes do we make to the livecheck code?
2. What tests do we write?
3. (Minor) Should livecheck be a common command (for completions)?

I will also be working on the RuboCops/audits in parallel. Looking forward to your reviews!

nandahkrishna commented:
Also, I'm unfamiliar with how the `man` change should be handled (that's why CI seems to failing). I ran `brew man` locally and it made some changes to the man files, if I should include those changes as well, let me know.

maxim-belkin commented:
> I ran `brew man` locally and it made some changes to the man files, if I should include those changes as well, let me know.

That's exactly what you should do.

nandahkrishna commented:
Thanks! Just had a look at the `brew license` PR as well and figured it out, should've done that before haha.

maxim-belkin commented:
I wonder if you could add people who contributed to the files that you add in this PR as co-authors: I think this is the right thing to do with such repo-to-repo migrations. You could do it file-by-file (that'll result in as many commits as there are files -- probably a lot) or once for the single commit.

nandahkrishna commented:
👍🏼 Totally agree. Maybe it's fine to add everyone as co-authors to the single commit, I think that keeps things cleaner in the PR. However, if you and the others think we should be more granular, I'll do just that.

maxim-belkin commented:
I don't have a preference -- this is just a way to acknowledge people's contributions on GitHub. Let's do it in a single commit for now.

dawidd6 commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -0,0 +1,464 @@
+
+require "cli/parser"
+
+require_relative "../livecheck_strategy"
```
Could this be:

```suggestion
require "lvecheck_strategy"
```

?

nandahkrishna replied:
Yeah that works, making the change 👍🏼

dawidd6 commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -0,0 +1,464 @@
+    end
+
+    if (cmd = args.named.first)
+      require?("livecheck/commands/#{cmd}") && return
```
I always wondered what does this line do, any explanations?

nandahkrishna replied:
There was a `brew livecheck cat` command before, to view a Livecheckable, analogous to `brew cat` to view the Formula source file. We removed that command a while back and this is no longer required (we can just use `brew cat` after the homebrew-core migration), I think. Thanks for pointing it out!

MikeMcQuaid commented on Library/Homebrew/livecheck_strategy.rb:
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module LivecheckStrategy
```
```suggestion
module Homebrew::Livecheck::Strategy
```

samford replied:
If strategies are moving to `livecheck/strategies` (e.g., `livecheck/strategies/apache`) should `livecheck_strategy.rb` be moved into the `livecheck` folder as well? If so, does the file need to be renamed?

MikeMcQuaid replied:
Yeh, I think so. `livecheck/strategy.rb` or similar may make sense.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -0,0 +1,459 @@
+    end
+  end
+
+  def livecheck
```
Would be good to have this function be as minimal as possible and essentially just parse/setup arguments and pass them to a method in the `Homebrew::Livecheck` module to actually do the main livecheck logic.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -0,0 +1,459 @@
+      return
+    end
+
+    # Identify any non-homebrew/core taps in use for current formulae
```
I'd say everything here and below should be in the `Homebrew::Livecheck` module.

samford replied:
With this in mind, we should probably move the constants that come before `def livecheck_args` (e.g., `WATCHLIST_PATH`, `GITHUB_SPECIAL_CASES`, etc.) into the `Homebrew::Livecheck` module as well, correct? If so, `@livecheck_strategy_names` could also be in `Homebrew::Livecheck`.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -0,0 +1,459 @@
+        end
+      end
+
+    if formulae_to_check.nil? || (formulae_to_check.is_a?(Array) && formulae_to_check.empty?)
+      onoe "No formulae to check."
+      return
+    end
```
```suggestion
    raise UsageError, "No formulae to check." if formulae_to_check.blank?
```

MikeMcQuaid commented on Library/Homebrew/livecheck_strategy/apache.rb:
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module LivecheckStrategy
+  class Apache
```
```suggestion
  module Homebrew::Livecheck::Strategies::Apache
```

and put it in `livecheck/strategies/apache` using `module_function` to avoid `self.`

samford replied:
We have `Homebrew::Livecheck::Strategy` above and `Homebrew::Livecheck::Strategies::Apache` here. Just to be sure, is the `Strategy` vs. `Strategies` mismatch intentional or should we be doing `Homebrew::Livecheck::Strategy::Apache` here?

MikeMcQuaid replied:
`Strategy` probably makes more sense to me.

samford replied:
I agree. It makes sense for `LivecheckStrategy` to become `Homebrew::Livecheck::Strategy` and for `LivecheckStrategy::Apache` to become `Homebrew::Livecheck::Strategy::Apache`.

Last question on this: It seems like we generally align the name of the Ruby file and the related directory (e.g., `version.rb`, `version`). Should we also use a `livecheck/strategy/` directory name, to align with `livecheck/strategy.rb` (as discussed in the other comment)?

MikeMcQuaid replied:
> Should we also use a `livecheck/strategy/` directory name

Yup 👍🏻

MikeMcQuaid commented on docs/Manpage.md:
```diff
@@ -880,6 +880,26 @@ provided, check all kegs. Raises an error if run on uninstalled formulae.
+Check for newer versions of formulae from upstream.
+
+If no formula argument is passed, the list of formulae to check is taken from
+`HOMEBREW_LIVECHECK_WATCHLIST` or `~/.brew_livecheck_watchlist`.
```
Add `HOMEBREW_LIVECHECK_WATCHLIST` to `Library/Homebrew/env_config.rb`

nandahkrishna commented:
The latest commit includes some refactoring done based on your suggestions @MikeMcQuaid.

I've added a `dev-cmd/livecheck_spec.rb` test similar to `linkage_spec`, but I'm not sure how exactly we'd have to proceed with testing `livecheck`'s core logic. There's a lot of network operations involved and actually testing a Formula might not be ideal as upstream can be unreliable.

How do you think we should approach tests for the strategies and livecheck 'core' code?

samford commented:
> There's a lot of network operations involved and actually testing a Formula might not be ideal as upstream can be unreliable.

I agree that a lot of livecheck involves network requests and this can make testing more difficult.

How do you feel about stubbing network requests, @MikeMcQuaid? It seems like we could use something like the [`webmock` gem](http://github.com/bblimke/webmock) to avoid external network requests (which can be slow and unreliable) and to control the response for the sake of testing.

MikeMcQuaid commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -0,0 +1,394 @@
+      "formula" => formula_name(formula, args: args),
+      "status"  => status_str,
+    }
+    status_hash["messages"] = messages if messages.is_a?(Array)
```
```suggestion
    status_hash["messages"] = messages.presence
```

MikeMcQuaid commented on Library/Homebrew/livecheck/strategy/apache.rb:
```diff
@@ -0,0 +1,18 @@
+  end
+
+  def find_versions(url, regex = nil)
+    path, prefix, suffix = url.match(%r{path=(.+?)/([^/]*?)\d+(?:\.\d+)+(/|[^/]*)})[1, 3]
```
This regex (and most of these strategies) is a bit inpeneratrable. You may want to use the multiline regex format, add comments, assign to a constant and explain what `1` and `3` refer to using named match groups instead.

nandahkrishna replied:
I've included this issue in the new PR for strategy migration (in the PR description), so I'll resolve it here.

MikeMcQuaid commented on Library/Homebrew/livecheck/strategy/git.rb:
```diff
@@ -0,0 +1,65 @@
+  PRIORITY = 8
+
+  def tag_info(repo_url, filter = nil)
+    stdout_str, stderr_str, _status = Open3.capture3(
```
Use `system_command` `Utils.popen_*` here instead.

MikeMcQuaid commented on Library/Homebrew/livecheck/strategy/git.rb:
```diff
@@ -0,0 +1,65 @@
+    )
+
+    tags_data = { tags: [] }
+    tags_data[:messages] = stderr_str.split("\n") unless stderr_str.empty?
```
Everywhere: prefer `blank?` over `empty?` or `nil?`

MikeMcQuaid commented on Library/Homebrew/livecheck/strategy/gnu.rb:
```diff
@@ -0,0 +1,45 @@
+  ].freeze
+  private_constant :PROJECT_NAME_REGEXES
+
+  SPECIAL_CASES = %w[
```
Would be nice to have a test that verifies all these formulae still exist in homebrew-core.

nandahkrishna replied:
@samford correct me if I'm wrong, but wouldn't the `strategy` feature make `SPECIAL_CASES` obsolete? Wouldn't we just have to use `strategy :page_match` or something to prevent the GNU strategy from being used?

samford replied:
Yes, for the `Gnu` and `Sourceforge` `SPECIAL_CASES` we can add/modify the `livecheck` block to use `strategy` instead. I was planning to address this after the Homebrew/brew migration but I can take care of the livecheckable changes if we want to do it as part of this process.

To be clear, the `Gnome` strategy’s `DEV_VERSION_ALLOWLIST` is different. This array constant is used in `#find_versions` to select the proper regex, instead of omitting the URL from matching the strategy.

MikeMcQuaid commented on Library/Homebrew/test/dev-cmd/livecheck_spec.rb:
```diff
@@ -0,0 +1,15 @@
+describe "brew livecheck", :integration_test do
+  it "errors when no Formula arguments are given and watchlist is empty" do
+    expect { brew "livecheck" }
+      .to output(/Invalid usage: No formulae to check./).to_stderr
```
Thoughts on having this check a single, specified formula?

nandahkrishna replied:
I was thinking of checking the `ruby` Formula, but since upstream is unreliable, crazy idea (I don't know how feasible it is though): how about a 'fake' temporary Formula that checks for Homebrew/brew's latest version or something?

MikeMcQuaid commented:
> I've added a `dev-cmd/livecheck_spec.rb` test similar to `linkage_spec`, but I'm not sure how exactly we'd have to proceed with testing `livecheck`'s core logic. There's a lot of network operations involved and actually testing a Formula might not be ideal as upstream can be unreliable.

I think ideally split out the methods based on whether they require a network connection or not. Those that do can have `:needs_network` tests and be really simple. 

> How do you think we should approach tests for the strategies and livecheck 'core' code?

For strategies I'd imagine it'd be a matter of being a bit like the versions tests; pass in some expected correct and incorrect URLs and ensure the strategy gets the expected results.

> How do you feel about stubbing network requests, @MikeMcQuaid? It seems like we could use something like the [`webmock` gem](https://github.com/bblimke/webmock) to avoid external network requests (which can be slow and unreliable) and to control the response for the sake of testing.

We've previously avoided this stubbing because, in my experience, your tests then become tests of the stubbing library more than tests of the actual code. To be clear: I don't expect 100% coverage here and if we can sufficiently extract network logic: that code can have very minimal (and in some cases, no) testing and save the more extensive tests for local logic.

samford commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -0,0 +1,72 @@
+        args.formulae
+      elsif File.exist?(Livecheck::WATCHLIST_PATH)
+        begin
+          File.readlines(Livecheck::WATCHLIST_PATH).each_with_object([]) do |word, memo|
+            memo << Formulary.factory(word.chomp) unless word.start_with?("#")
+          end
```
Was `Formula[line]` meant here instead of `Formula[word]`?

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -0,0 +1,394 @@
+  def formula_name(formula, args:)
+    args.full_name? ? formula.full_name : formula.name
+  end
+  private_class_method :formula_name
```
Would anywhere mean methods in `livecheck/strategy.rb` or individual strategies as well? Just confirming because we have a `private_class_method :strategies` in `unpack_strategy.rb`.

nandahkrishna commented:
With the latest push, this PR will only deal with:
* Migrating the `livecheck` command as a `dev-cmd` to Homebrew/brew.
* Tests for the `dev-cmd`.
* Completions for `livecheck`.

I will be creating PRs for the migration of the `livecheck` module and the strategies. This PR depends on those two, and cannot be merged until they have been merged. I'll reference them here after I create them.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -0,0 +1,76 @@
+    args = livecheck_args.parse
+
+    if args.debug? && args.verbose?
+      puts ARGV.inspect
+      puts args
```
Is it worth having both? Should this have a header?

samford replied:
> Is it worth having both?

It looks like [you added `args` to this output](https://github.com/Homebrew/homebrew-livecheck/commit/3d63817fa3974d4ed5594141dfe6d0a2e5588c5e), so I would say it's up to you.

I imagine it would only be useful to have both if/when the `args` parsing isn't working properly. However, we can always add temporary output like this when we're debugging issues, so there may not be much value in the interim time.

If it were up to me, I would probably favor `args` or even removing these two output lines entirely if there aren't other `brew` commands that do the same thing.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -0,0 +1,76 @@
+      puts ENV["HOMEBREW_LIVECHECK_WATCHLIST"] if ENV["HOMEBREW_LIVECHECK_WATCHLIST"].present?
+    end
+
+    formulae_to_check =
+      if args.tap
+        Tap.fetch(args.tap).formula_names.map { |name| Formula[name] }
+      elsif args.installed?
```
```suggestion
    formulae_to_check = if args.tap
        Tap.fetch(args.tap).formula_names.map { |name| Formula[name] }
    elsif args.installed?
```

MikeMcQuaid commented on Library/Homebrew/test/dev-cmd/livecheck_spec.rb:
```diff
@@ -0,0 +1,29 @@
+      .and not_to_output.to_stdout
+  end
+
+  it "reports the latest version of a Formula", :needs_network do
```
Let's limit these to a single test (probably this one)

nandahkrishna commented:
The following changes have been made in the latest commit:
* Removed `ARGV.inspect`, retaining only `args`. I couldn't find a command that prints the args but `pr-upload.rb` contains something similar during a dry-run. I decided that it would be okay to just leave the `puts args` there (it's under the `-d` and `-v` switches anyway), unless @MikeMcQuaid feels it isn't necessary.
* Retained only one test (which runs livecheck for a test formula).
* Fixed the `if` style issue.

MikeMcQuaid commented:
Looks good! Happy for this to merge when CI is 💚. Great work @nandahkrishna!

samford commented:
This relies on #8254 and 8255 to work, so we probably don't want to merge this PR until after those other two are finished and merged, correct?

MikeMcQuaid commented:
Agreed.

nandahkrishna commented:
Incorporating the suggestion and rebasing.

MikeMcQuaid reviewed (commented):
Looks good to me. We probably want to ensure we delete the existing https://github.com/homebrew/homebrew-livecheck command and library code when we merge this and archive that repository. It'd also be worth adding to `DEPRECATED_OFFICIAL_TAPS` in `Library/Homebrew/official_taps.rb` in this PR, too!

nandahkrishna commented:
> It'd also be worth adding to `DEPRECATED_OFFICIAL_TAPS` in `Library/Homebrew/official_taps.rb` in this PR, too!
Added this 👍🏼

Will rebase and push after #8544 is merged.

MikeMcQuaid commented:
> We probably want to ensure we delete the existing https://github.com/homebrew/homebrew-livecheck command and library code when we merge this

@nandahkrishna Could you open a PR for this, too? I'll merge it when this is merged.

nandahkrishna commented:
Sure Mike 👍🏼

MikeMcQuaid commented:
Thanks again @nandahkrishna!
